File: Longest_palindrom_substring.py
'''
Given a string s, find the longest palindromic substring in s.
Input: "babad"
Output: "bab"
Note: "aba" is also a valid answer.
https://leetcode.com/problems/longest-palindromic-substring/
'''


def longestPalindrome(s: str) -> str:
    if s == "":
        return ""
    else:
        max_string = ""
        for i in range(0, len(s) - 1):
            for j in range(len(s), i + 1, -1):
                if s[i:j] == s[i:j][::-1]:
                    if len(max_string) < len(s[i:j]):
                        max_string = s[i:j]
        if max_string == "":
            max_string = s[0]
        return max_string


if __name__ == '__main__':
    print(longestPalindrome("abacda"))

File: Longest_palindrom_substring.java

/*
Given a string s, find the longest palindromic substring in s.
Input: "babad"
Output: "bab"
Note: "aba" is also a valid answer.
https://leetcode.com/problems/longest-palindromic-substring/
 */
import java.util.*;
import java.util.stream.*;

class Solution {
	public static String longestPalindrome(String s) {
		String max_string = "";
		for (int i = 0; i < s.length(); i++) {
			String s1 = extend(s, i, i), s2 = extend(s, i, i + 1);
			if (s1.length() > max_string.length()) {
				max_string = s1;
			}
			if (s2.length() > max_string.length()) {
				max_string = s2;
			}
		}
		return max_string;
	}

	private static String extend(String s, int i, int j) {
		try {
			for (; 0 <= i && j < s.length(); i--, j++) {
				if (s.charAt(i) != s.charAt(j)) {
					break;
				}
			}
		} catch (Exception e) {
			System.out.print(e);
		}
		return s.substring(i + 1, j);
	}

	public static void main(String[] args) {
		System.out.print(longestPalindrome("cbbd"));
	}
}
